crates/shared/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -388,13 +389,26 @@ impl PoolFetching for UniswapV3PoolFetcher {
 }
 
 /// For a given checkpoint, append events to get a new checkpoint
-fn append_events(pools: &mut HashMap<H160, PoolInfo>, events: Vec<Event<UniswapV3Event>>) {
+fn append_events(pools: &mut HashMap<H160, Arc<PoolInfo>>, events: Vec<Event<UniswapV3Event>>) {
     for event in events {
         let address = event
             .meta
             .expect("metadata must exist for mined blocks")
             .address;
-        if let Some(pool) = pools.get_mut(&address).map(|pool| &mut pool.state) {
+
+        if let Entry::Occupied(mut entry) = pools.entry(address) {
+            // Because contents of `Arc`s can't be modified we:
+            // 1. swap out the current state with a dummy
+            // 2. update current state
+            // 3. put updated state into `Arc` and swap back into the cache
+            let mut current = {
```

Comment:
I see the caller of this functions is an `async` function. Why not to use a tokio `Mutex`, `RwLock`, `parking_lot` or other solution to do this write? `Arc` can be accessed by multiple readers, and create a race conditions. You can argue that you are very sure this is not accessed multiple times by many readers, but then it forces new developers or developers who don't know this to be aware of it, if the functionality were to be extended (let's say we add another parallel function for the maintenance of Uniswap, etc)

Reply:
> Arc can be accessed by multiple readers, and create a race conditions.

Arc can be accessed by multiple readers but it can not be modified which is why you can't cause race conditions with it. That is precisely the reason I have to jump through these hoops. The only thing that can happen is that:
1. somebody fetches liquidity (by cloning the pool)
2. I update the cache
3. they don't see the update
However, this is already the case so in that sense from the perspective of an outside observer nothing about this function changed (Except that it's now returning the data inside an `Arc`).

> You can argue that you are very sure this is not accessed multiple times by many readers, but then it forces new developers or developers who don't know this to be aware of it

No, other developers don't have to be careful about anything here.
The state I'm modifying is guaranteed to be accessed only by me during the update.
First of all we get an exclusive reference to the HashMap so we know for sure nobody else is modifying that map at the same time.
Secondly because I use `Arc::unwrap_or_clone()` either nobody else is holding a copy of that particular `Arc` (then it's safe to unwrap) or somebody is currently using it then we simply clone and modify the cloned version. Modifying the cloned version is also safe because nobody can get access to it until I'm done with it and put it back into the map.

> Why not to use a tokio Mutex, RwLock, parking_lot or other solution to do this write?

They don't really help here. The thing I'm trying to solve is expensive copies of the pool state. That can only be solved with some sort of reference counting. So the only thing I could be doing is to use `Mutex<Arc<Data>>`. However, since we already have an exclusive access to the map the `Mutex` is not needed to protect against race conditions so it would just be overhead.

Reply:
> Arc can be accessed by multiple readers but it can not be modified which is why you can't cause race conditions with it.

This is not correct, you are swapping the memory of the `Arc`, therefore modifying its content in order to modify it, and substituting the `Arc` pointing to a `dummy` which is not even containing an object the same structure as `PoolInfo`. Making any access by another party to that `Arc` extremely dangeous.

> The state I'm modifying is guaranteed to be accessed only by me during the update.
First of all we get an exclusive reference to the HashMap so we know for sure nobody else is modifying that map at the same time.

This is partially true. Because you are passing `&mut HashMap<>` doesn't mean that someone has references of the `Arc`s inside the `HashMap` (that's the whole point of the `Arc` after all). Indeed you are making the `HashMap` mutable only for your function, but the access to the `Arc` is still shared among everyone who wants to clone it.

> Secondly because I use Arc::unwrap_or_clone() either nobody else is holding a copy of that particular Arc (then it's safe to unwrap) or somebody is currently using it then we simply clone and modify the cloned version. Modifying the cloned version is also safe because nobody can get access to it until I'm done with it and put it back into the map.

This is true, but this only guarantees you that the data you modify is either not used by someone else, or in that case, it is clone. But we can have this scenario:

Someone else clones the `Arc` before that function. If another tokio thread/task cloned the entry Arc before this code executes, it retains a separate reference to the original data. Swapping out the contents of the entry Arc with a dummy value does not affect this separate reference. If the other thread concurrently accesses or modifies the data through its reference to the entry `Arc`, it can lead to a race condition. This can definitely result in unpredictable behavior, data corruption, or crashes.

> They don't really help here. The thing I'm trying to solve is expensive copies of the pool state. That can only be solved with some sort of reference counting. So the only thing I could be doing is to use Mutex<Arc<Data>>. However, since we already have an exclusive access to the map the Mutex is not needed to protect against race conditions so it would just be overhead.

It should be the opposite `Arc<Mutex<>>` :thinking:  But @MartinquaXD , we have exclusive access to the map, but we don't have exclusive access to the `Arc` contained in the map :pray:

Reply:
We resolved the race condition concerns in a call. It's correct that using `RwLock` around each pool might be faster overall depending on how often we actually try to update the state while somebody else still has access to the data. I will add metrics to see how often that actually happens. The more often it happens the more likely using `RwLock` would be the right choice here.